### DISCUSSION TITLE
Change a component without triggering change detection

### DISCUSSION THREAD

**Discussion by raffaeleragni** [bevyengine/bevy#8748] [Q&A] [answered]
Hi, I have a peculiar use case where I am synchronizing data across the network, and my intention is to base these synchronizations into ordered events that are not aware of game ticks or any other specific engine detail.
Naturally, when I combine this network synchronization, that updates component's data, with change detection in order to trigger such updates, I end up in a loop. Currently I am checking if the values are really changed and if not then I won't update the component's data, but that still makes my network packets go roundtrip at least once.

All of that is made more complicated on the fact that I need to use reflection because these may be future components implemented in a higher level crate, that is using my sync one as a dependency. This means I am bound to use the reflection apis that are available in bevy.

So, is there a way to apply a component change without triggering the change in a system's query that is looking for such trigger?

Currently I am using this way of updating components that came across the net:
```rust
reflect_component.apply_or_insert(&mut world.entity_mut(e_id), component_data.as_reflect());
```
(ref: https://github.com/raffaeleragni/bevy_sync/blob/master/src/server.rs#L409)

**Comment by alice-i-cecile** [comment]
You want the [`bypass_change_detection`](https://docs.rs/bevy/latest/bevy/ecs/change_detection/trait.DetectChangesMut.html#tymethod.bypass_change_detection) method, which should work for both components and resources. If it's not cooperating for a reflection based workflow, please open an issue!

**Comment by raffaeleragni** [comment]
Yes, I noticed that in a PR before, but that's also the reason I am asking this question: I can't find how I am supposed to use it.

**Reply by rlidwka** [reply, marked as answer]
> Yes, I noticed that in a PR before, but that's also the reason I am asking this question: I can't find how I am supposed to use it.

Instead of `&mut component` you use `component.bypass_change_detection()`.

> Naturally, when I combine this network synchronization, that updates component's data, with change detection in order to trigger such updates, I end up in a loop.

If you put both sides of the sync (bevy to network and network to bevy) into the same system, you won't have any loops. This is my preferred approach at the moment.

**Comment by rlidwka** [comment]
> If you put both sides of the sync (bevy to network and network to bevy) into the same system, you won't have any loops. This is my preferred approach at the moment.

Speaking of which, @alice-i-cecile is there a way to have two systems that would ignore changes made by each other (i.e. share state where changes are tracked)?

**Reply by alice-i-cecile** [reply]
Hmm. I don't know of any particularly good ways to do that. You could probably do terrible things by overriding the system's change ticks to match the other, but that's not going to be particularly robust or pretty.

**Comment by raffaeleragni** [comment]
For the moment I can do this:
```rust
if let Some(mut component) = reflect_component.reflect_mut(entity) {
    component.bypass_change_detection().apply(component_data.as_reflect());
} else {
    reflect_component.insert(entity, component_data.as_reflect());
}
```

In the else statement tho, I still need to pass through reflection, maybe I just need to figure out how to do that in a different way too, since entities are also involved for that case and not just components (EntityMut::insert is used in side ReflectComponent::insert).
